Extraneous interaction regions for pseudo elements and some composited childrens

etiennesegonzac opened the pull request:
#### ab29cfa512af8fac599ae1f0e7bb19b99980cc52
<pre>
Extraneous interaction regions for pseudo elements and some composited childrens
<a href="https://bugs.webkit.org/show_bug.cgi?id=251995">https://bugs.webkit.org/show_bug.cgi?id=251995</a>
&lt;rdar://104140617&gt;

Reviewed by Tim Horton.

Tweak the interaction regions generation heuristics to avoid extra
regions for:
- pseudo-elements
- children of block elements getting their own InteractionRegion

* Source/WebCore/page/InteractionRegion.cpp:
(WebCore::interactionRegionForRenderedRegion):
Keep track of the original RegionRenderer node separately of the one we
end up matching in order to tweak the heuristics.

* LayoutTests/interaction-region/nested-composited-text-painter-expected.txt: Added.
* LayoutTests/interaction-region/nested-composited-text-painter.html: Added.
* LayoutTests/interaction-region/pseudo-element-expected.txt: Added.
* LayoutTests/interaction-region/pseudo-element.html: Added.
Add tests for these cases.

Canonical link: <a href="https://commits.webkit.org/260567@main">https://commits.webkit.org/260567@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7a2a5e94e7999f39cd680cbdb299ce47b6ce12dc

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/108640 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/17741 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/41494 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/117750 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/117951 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/112523 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/19192 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/9018 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/100872 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/114407 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/14389 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/97622 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/42358 "Found 2 new test failures: webgl/2.0.0/conformance2/textures/image_bitmap_from_video/tex-2d-r8ui-red_integer-unsigned_byte.html, webgl/2.0.0/conformance2/textures/image_data/tex-2d-r8-red-unsigned_byte.html (failure)") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/96365 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/29260 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/84088 "Found 3 new API test failures: /TestWebKit:WebKit.OnDeviceChangeCrash, /TestWebKit:WebKit.PrivateBrowsingPushStateNoHistoryCallback, /WebKitGTK/TestWebViewEditor:/webkit/WebKitWebView/cut-copy-paste/editable (failure)") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/10549 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/30610 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/11307 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/7524 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/16698 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/50206 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7307 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/12895 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

mwyrzykowski reviewed (commented):
Looks good to me, thank you for fixing this. I will defer approval to @hortont424

etiennesegonzac committed:
Extraneous interaction regions for pseudo elements and some composited childrens
https://bugs.webkit.org/show_bug.cgi?id=251995
<rdar://104140617>

Reviewed by Tim Horton.

Tweak the interaction regions generation heuristics to avoid extra
regions for:
- pseudo-elements
- children of block elements getting their own InteractionRegion

* Source/WebCore/page/InteractionRegion.cpp:
(WebCore::interactionRegionForRenderedRegion):
Keep track of the original RegionRenderer node separately of the one we
end up matching in order to tweak the heuristics.

* LayoutTests/interaction-region/nested-composited-text-painter-expected.txt: Added.
* LayoutTests/interaction-region/nested-composited-text-painter.html: Added.
* LayoutTests/interaction-region/pseudo-element-expected.txt: Added.
* LayoutTests/interaction-region/pseudo-element.html: Added.
Add tests for these cases.

Canonical link: https://commits.webkit.org/260567@main
